### ISSUE TITLE
Missing default ObjectCodec when decoding from XML

### ISSUE THREAD

**Issue by zwlxt** [maxGraph/maxgraph#178]
Version: release 0.1.0

I was trying to decode the following XML:

```XML
<GraphDataModel><root><Cell id="0"><Object as="style"/></Cell><Cell id="1" parent="0"><Object as="style"/></Cell><Cell id="B_#0" value="rootNode" vertex="1" parent="1"><Geometry _x="100" _y="100" _width="100" _height="80" as="geometry"/><Object as="style"/></Cell></root></GraphDataModel>
```

```javascript
const doc = xmlUtils.parseXml(str);
const codec = new Codec(doc);
codec.decode(doc.documentElement, this.graph.getDataModel());
```

And got error like this:

```
Uncaught Error: Invalid x supplied.
    at set x [as x] (Point.js:44:19)
    at GraphView.updateCellState (GraphView.js:743:32)
    at GraphView.validateCellState (GraphView.js:676:26)
    at GraphView.validateCellState (GraphView.js:691:30)
    at GraphView.validateCellState (GraphView.js:691:30)
    at GraphView.validate (GraphView.js:444:58)
    at ModelEditorGraph.graphModelChanged (Graph.js:429:19)
    at Graph.graphModelChangeListener (Graph.js:348:18)
    at GraphDataModel.fireEvent (EventSource.js:132:41)
    at edit.notify (GraphDataModel.js:958:25)
```

After doing a lot of debugging, I found that `CodecRegistry.getCodecByName(node.nodeName)` does not return the default ObjectCodec for some elements (like `Point`, `Geometry`, etc... ). Therefore the decoding procedure returns the original xml element, and breaks later procedure. 

I also found that if I do encoding before decoding, no error will occurs because encoding procedure uses `CodecRegistry.getCodec(constructor_)` which `Registers a new default codec for the given constructor, if no codec has been previously defined.` 

My possible workaround: Manually register ObjectCodec for missing codecs:
```javascript
CodecRegistry.register(new ObjectCodec(new Geometry()));
CodecRegistry.register(new ObjectCodec(new Point()));
CodecRegistry.register(new ObjectCodec(new Array()));
```

Perhaps `CodecRegistry.getCodecByName(node.nodeName)` should return a default codec like `CodecRegistry.getCodec(constructor_)` does?

**Comment by tbouffard** [comment]
Hello and thank you for these very detailed feedback and investigations 💯 !

First of all, please be aware that this feature is not mature as it was not reworked during the migration from `mxGraph`.
There are several other issues that have been filled on XML encoding/decoding, for example #102 and #169. Some corrections have been made to #125.

As I said in #169, there are currently no working examples that show this functionality (they probably won't work because of the problem you noticed). From my side, I consider this a perfect topic to add more automatic tests (#153) because it is a typical input/output flow: from string to model and from model to string.

There is also a lot to be done to improve the way codecs are stored. Currently, they are stored globally and they limit the tree shaking capabilities. Not to mention that you detect that decoding works if encoding is done first. What a side effect!

So, what I can propose is to first set up the infrastructure for automatic testing and a new class that will handle the import (from the chain) and export (to the chain) of the model. Then, new contributions will be easier to fix all the problems already detected by users.

Does this make sense to you?
Will you have time to contribute in the future to this feature (Pull Request, tests, review, ...)?

**Comment by tbouffard** [comment]
@zwlxt I have just created #183 which replicates most of the problems you have described here.

**Comment by tbouffard** [comment]
Possible workaround, as implemented in #297
```ts
CodecRegistry.register(new ObjectCodec(new Geometry()));
CodecRegistry.register(new ObjectCodec(new Point()));
CodecRegistry.register(new ObjectCodec({})); // Object
CodecRegistry.register(new ObjectCodec([])); // Array
```
